### ISSUE TITLE
Blazor WASM does not work in Edge Enhanced Security mode

### ISSUE THREAD

**Issue by Andrzej-W** [dotnet/aspnetcore#42641]
It's not a Blazor issue, but I think it should be documented and preferably resolved by the Edge browser team.

Usually I use Firefox, sometimes Chrome, but recently I decided to check the Blazor WASM application in the Edge browser. Unfortunately I have noticed this message in the browser console window: **Uncaught Error: This browser does not support WebAssembly.**

What??? Is Microsoft going to kill Blazor?

I have Windows 10 21H2 with all updates and Edge 103.0.1264.49. You have to download WASM application from the Internet - problem is not visible when you test locally.  You can try this: https://blazorstrap.io/V5/

To make it working you have to disable "security mitigations" completely. Changing from Strict to Balanced is not enough.

![obraz](https://user-images.githubusercontent.com/32604409/178127502-4a42628f-ce09-4672-a0c0-2aae2ea3d982.png)

I have no idea if this setting is enabled by default, or if I turned it on manually at some point.

After more than 4 years we have to much problems with Blazor WASM: it is big, it is slow, it is blocked by firewalls/antivirus software, and now it doesn't work in Edge.

Best wishes to the Blazor team. I still hope Blazor will revolutionize web application development.

**Comment by SteveSandersonMS** [comment]
Thanks @Andrzej-W. We are aware of this issue and have been in discussions with the Edge browser team. As far as I know:

 * This setting is not enabled by default, and is only a beta-quality feature currently (for example, the fact that it disables WebAssembly is obviously absurd). You have to turn it on manually, though maybe there's some way for sysadmins to turn it on automatically through group policy or similar.
 * The Edge team has implemented support for WebAssembly in an upcoming version of this feature. I don't have an exact ETA.

We are still working with the Edge team on trying to ensure that it doesn't have an adverse perf impact on WebAssembly apps.

cc @danroth27 in case he has any further information.

**Comment by danroth27** [comment]
As @SteveSandersonMS mentioned, ESM will support WebAssembly going forward. We are working with the Edge team to ensure   Blazor scenarios are not negatively impacted.

**Comment by Jelgab** [comment]
This ticket shows up as closed. Does it mean this issue has already been fixed? Supposedly, I have the lates version of Edge and I got that same error today (While I was imparting a live training on Blazor and decided to switch from Chrome to Edge to show that it works on all browsers. Unbelievable). After the training, I had to totally disable "Enhance your security on the web" to make it work.


![2022-07-14_12-57-24 - Enhance your security on the web - Turned on - Blazor error](https://user-images.githubusercontent.com/1036858/179042055-db26066e-869c-492c-9fe0-a3cd7acc49da.png)

![2022-07-14_12-57-24 - Enhance your security on the web - Turned off](https://user-images.githubusercontent.com/1036858/179040848-49751d4b-afc4-4909-9561-3df2e18ba97d.png)

**Comment by danroth27** [comment]
@Jelgab The Edge team is adding WebAssembly support to Enhanced Security Mode (ESM) via a [WebAssembly interpreter ("DrumBrake")](https://microsoftedge.github.io/edgevr/posts/Introducing-Enhanced-Security-for-Microsoft-Edge/) so that WebAssembly scenarios don't just break. However, performance is still a concern, and we are in discussions with the Edge team on that topic. Since we don't directly own this Edge behavior, please share your feedback about ESM directly with the Edge team.

**Comment by Jelgab** [comment]
@danroth27, thank you for the update. I did a quick read on Drum Brake and it looks like this WASM interpreter will be out very soon. That is good news!

**Comment by Andrzej-W** [comment]
@danroth27, @SteveSandersonMS, I'm really worried about the future of Blazor WASM. It was initially advertised as a tool to write web applications with near native performance. Now we know that it is much slower than native code, especially when it runs in the interpreter. I have read article about DrumBrake and found this information:
https://docs.google.com/document/d/1OIJ4Sv2XfTlI5NmTS1QI8v8wPL0LUT5s1W2D9OlJmMc/preview#heading=h.8enk175z61od
In short: WASM application interpreted by DrumBrake is 11.5 times slower than application compiled by Liftoff + Turbofan.
Blazor application compiled to IL language will by interpreted by WASM interpreter which will be interpreted by DrumBrake interpreter. It will be slow like a turtle.

**Comment by danroth27** [comment]
@Andrzej-W Yup, we're aware. This doesn't just impact Blazor WebAssembly - it potentially impacts any WebAssembly based site. We're actively discussing this with the Edge folks.

**Comment by christopherdude** [comment]
Would this comment above warrant keeping this ticket open for the time being?  This is a very large issue for a lot of developers as well as companies who are anywhere near the decision making point of using WASM.

**Comment by SteveSandersonMS** [comment]
@christopherdude Yes, we'll keep it open as things are still being explored.
